openverse_catalog/dags/common/loader/reporting.py

```diff
@@ -28,21 +32,54 @@ def humanize_time_duration(seconds):
     return ", ".join(parts)
 
 
-def report_completion(provider_name, duration, record_counts_by_media_type):
+def report_completion(
+    provider_name: str,
+    duration: float | str | None,
+    record_counts_by_media_type: RecordCounts,
+) -> None:
     """
     Send a Slack notification when the load_data task has completed.
     Messages are only sent out in production and if a Slack connection is defined.
     In all cases the data is logged.
+
+    The following data is reported:
+        - `duration`: The time the pull data task took to complete. This value is
+          "No data" in cases where the pull data task failed.
+        - `cleaned`: The number of records that were removed after the data was loaded
+          into a temporary table. This could be the result of missing columns, or
+          due to multiple records that have the same provider & foreign ID.
+        - `duplicates`: The number of records that have unique provider & foreign IDs,
+          but are duplicated across URL. This can occur when a provider makes multiple,
+          discrete references to the same source media within their API.
+        - `upserted`: The final number of records that made it into the media table
+          within the catalog database.
     """
     # Truncate the duration value if it's provided
     if isinstance(duration, float):
         duration = humanize_time_duration(duration)
 
     # List record count per media type
-    media_type_reports = "\n".join(
-        f"  - `{media_type}`: {record_count or '_No data_'}"
-        for media_type, record_count in record_counts_by_media_type.items()
-    )
+    media_type_reports = ""
+    for media_type, counts in record_counts_by_media_type.items():
+        loaded, cleaned, upserted = counts
+        if not upserted:
+            upserted_human_readable = "_No data_"
+        else:
+            upserted_human_readable = f"{upserted:,}"
+        media_type_reports += f"  - `{media_type}`: {upserted_human_readable}"
+        if any([count is None for count in counts]):
+            # Can't make calculation without data
+            continue
+        duplicates = loaded - cleaned - upserted
+        if duplicates or cleaned:
+            extras = []
+            if cleaned:
+                extras.append(f"{cleaned:,} cleaned")
+            if duplicates:
+                extras.append(f"{duplicates:,} duplicates")
+            if joined := ", ".join(extras):
+                media_type_reports += f" _({joined})_"
```

Comment:
nit: This packs a lot of logic into this little line! Took me a while to understand that it works. To be fair, I think I'm one of the poor souls that does _not_ like the walrus operator most of the time... but in this case it seems like too much going on, splitting it like this would be the same number of lines but much easier to read, at least in my opinion. Feel free to take it or leave it though!

```suggestion
            if extras:
                media_type_reports += f" _({', '.join(extras)})_"
```

Reply:
Oh, this is much better! I think I was avoiding the `join` call within the f-string because we so frequently have problems with escape characters not being allowed in f-strings.